Use phply for parsing PHP files

nijel opened the pull request:
Current PHP parser is really problematic, so how about using some real parser for PHP?

I've tried to base on https://github.com/viraptor/phply. ~~There are some tricky parts to solve (naming of units is slightly different and comments extraction does not work), but I'd like to get some feedback before spending more time on this.~~ It tries to be as much compatible as possible, but I've decided to change comments handling, see https://github.com/translate/translate/pull/3697#issuecomment-332743699 for more details.

This bases on https://github.com/translate/translate/pull/3696 to get fixed some errors in tests.

Fixes #1685
Fixes #1898
Fixes #1929
Fixes #2611
Fixes #2612
Fixes #2647
Fixes #2648
Fixes #3608
Fixes #3625
Fixes #3626
Fixes #3627
Fixes #3628
Fixes #3629

nijel commented:
In the end the remaining bits were not that complex, so the PR is complete now.

It does change handling of comments - now it's consistently including comment markers (previously it did so for `/* ... */` comments but did not for `//...` comments) and multiline block comments are inserted as one note instead of splitting them up. Both could be changed, but I think this way it's better.

nijel commented:
The Python 2 tests will fail until https://github.com/viraptor/phply/pull/33 is accepted and released.

nijel commented:
Okay, the above mentioned fix got merged, but new version introduced another regression, so the tests will fail until https://github.com/viraptor/phply/pull/39 is merged...

nijel commented:
All needed fixes have landed in phply, so all tests now pass :-).

nijel commented:
BTW: This is running since September 2017 on Hosted Weblate without any problems, so at least for storage part it is pretty well tested in the real life.

unho commented on translate/storage/php.py:
```diff
@@ -54,13 +54,133 @@
 import six
+from phply.phpparse import make_parser
+from phply.phplex import lexer, FilteredLexer, full_lexer, unparsed
+from phply.phpast import InlineHTML, FunctionCall, Assignment, ArrayOffset, Variable, Array, ArrayElement, Return, BinaryOp, Node
```
These imports don't follow our import guidelines: http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#imports

Also the lines shouldn't be that long.

phlax replied:
wierd that it didnt break any linting checks @unho

unho commented on translate/storage/php.py:
```diff
@@ -54,13 +54,133 @@
+
+    def token(self):
+        # Copied from FilteredLexer, only difference is that it uses
+        # next_token method
```
Nit: If the comment applies to the whole method, then make it a docstring.

unho commented on translate/storage/test_php.py:
```diff
@@ -164,7 +165,7 @@ def test_comment_output(self):
  * Comment line 2
  */
-$foo='bar';
+$foo = 'bar';
```
Remove the comment 5 lines up since I think no longer applies.

unho commented on translate/storage/test_php.py:
```diff
@@ -763,3 +755,8 @@ def test_simpledefinition_after_define(self):
         assert phpunit.source == "Bestand selectie"
+
+    def test_quotes(self):
+        phpsource = b'<?php\n$txt[\'DISPLAYEDINFOS\'] = "<a href=\\"__PARAM2__\\">Modificar...</a>";\n'
```
Is it not possible to use triple quoted string to avoid having `\n` in the middle of the string?

unho commented on translate/storage/test_php.py:
```diff
@@ -762,6 +762,17 @@ def test_quotes(self):
+        phpsource = """
+$lang['mediaselect'] = "Really \\\"something\\\"" . $variable . "\\\"something else\\\"";"""
+        phpfile = self.phpparse(phpsource)
+        print(len(phpfile.units))
```
Unnecessary `print` call.

nijel replied:
It's used all over the file this way, so it's there after copying existing code. Will remove that in separate commit.

unho replied:
Thanks.

unho commented:
Can you please add a specific test for #1898?

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
+        def concatenate(item):
+            if isinstance(item, six.string_types):
+                return item
+            assert isinstance(item, BinaryOp)
```
Is this `assert` necessary? If not please remove.

nijel replied:
It's there merely to catch in case parser produces something wrong. Therefore I've used assert which would discover this while running tests, but it's situation which really should not happen.

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
+
+        def handle_array(prefix, nodes, lexer):
+            for item in nodes:
+                assert isinstance(item, ArrayElement)
```
Again please avoid use of `assert` unless necessary.

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
-            # previous translation lines.
-            if invalue:
-                lastvalue = lastvalue + value + "\n"
+        def newunit(name, value, escape_type, comments):
```
I would rename this to `create_and_add_unit` and make it a method.

nijel replied:
Good point. At some earlier point it did use some of the namespace, but it's now gone.

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
-            if invalue:
-                lastvalue = lastvalue + value + "\n"
+        def newunit(name, value, escape_type, comments):
+            newunit = phpunit()
```
Nit: Use `self.UnitClass()` instead of `phpunit()`.

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
+        for item in parser.productions:
+            item.callable = wrap_production(item.callable)
+        lexer = PHPLexer()
+        pos = 0
```
Unused variable?

unho commented on translate/storage/php.py:
```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
+        tree = parser.parse(phpsrc.decode(self.encoding), lexer=lexer, tracking=True)
+        # Handle text without PHP start
+        if len(tree) == 1 and isinstance(tree[0], InlineHTML):
+            return self.parse(b'<?php\n' + phpsrc)
```
Is it necessary to call `self.parse`? Doesn't the following work?:

```python
            phpsrc = b'<?php\n' + phpsrc
            tree = parser.parse(phpsrc.decode(self.encoding), lexer=lexer, tracking=True)
```

This way we should be able to avoid initializing stuff again.

nijel replied:
The lexer has internal state, so it should be reinitialized. I'm not sure about parser, so I've rather did it the safe way...

unho replied:
I am ok with leaving it as is.

unho commented on translate/storage/php.py:
```diff
@@ -54,13 +54,133 @@
+                else:
+                    # Rewrite close tag as a semicolon.
+                    t.type = 'SEMI'
+                    break
```
Better to shorten this code block like:

```python
            # Insert semicolons in place of close tags where necessary.
            avoid_tags = ('OPEN_TAG', 'SEMI', 'COLON', 'LBRACE', 'RBRACE')
            rewrite_as_semicolon = (t.type == 'CLOSE_TAG' and
                                    (not self.last_token or
                                     self.last_token.type not in avoid_tags))
            if rewrite_as_semicolon:
                # Rewrite close tag as a semicolon.
                t.type = 'SEMI'
                break
```

unho commented:
@nijel This is really amazing. I've left some comments and nits. I am really **really** sorry about it took me so long to review this great contribution. lgtm once comments are addressed.

unho commented:
Weird, some of the comments I've made are shown collapsed despite they still apply.

nijel commented:
Thanks for the review, I've addressed most of the comments (though it's possible that I've missed something due to GitHub hiding them as outdated). The only thing which I didn't change are the assert statements, I think they are useful.

unho commented on translate/storage/php.py:
```diff
@@ -54,13 +54,140 @@
+from phply.phplex import lexer, FilteredLexer, full_lexer, unparsed
+from phply.phpast import (InlineHTML, FunctionCall, Assignment, ArrayOffset,
+                          Variable, Array, ArrayElement, Return, BinaryOp,
+                          Node)
```
These imports still need some changes to conform to http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#imports

unho commented on translate/storage/test_php.py:
```diff
@@ -755,11 +733,131 @@ def test_simpledefinition_after_define(self):
+        assert phpfile.__bytes__() == phpsource
+
+    def test_space_before_comma(self):
+        """check that spacing before comma or semicolon doesn't break parser"""
```
Nit: Add reference in docstring to issue #1898

unho commented on translate/storage/test_php.py:
```diff
@@ -755,11 +733,131 @@ def test_simpledefinition_after_define(self):
+        );
+        $month_jan = 'Jan';
+        $month_feb = 'Feb'  ;
+        $month_mar = 'Mar';
```
Also include the following test cases:

```php
define("_SEARCH","Search");
define("_LOGIN","Login"  );
define("_POLLS","Polls");
```

unho commented on translate/storage/test_php.py:
```diff
@@ -755,11 +733,131 @@ def test_simpledefinition_after_define(self):
+        assert len(phpfile.units) == 6
+
+    def test_equals_in_id(self):
+        """check that equals in id doesn't break parser"""
```
Please add reference to issue #1929 in test's docstring.

unho commented on translate/storage/test_php.py:
```diff
@@ -755,11 +733,131 @@ def test_simpledefinition_after_define(self):
+        assert phpfile.units[0].name == "$strings['key = value']"
+
+    def test_comma_in_string(self):
+        """check that comma in string doesn't break parser"""
```
Also add reference in this docstring to the related issue #3608.

unho commented on translate/storage/test_php.py:
```diff
@@ -755,11 +733,131 @@ def test_simpledefinition_after_define(self):
+        assert phpfile.units[0].name == '$str'
+
+    def test_concatenation(self):
+        """check parsing concatenadted strings"""
```
Typo: concatenadted

unho commented:
@nijel reviewed the PR again and raised some extra nits, mostly on the new changes. lgtm after those are addressed. Again, thanks a lot!

unho commented:
Sorry, hit the wrong button.

nijel commented:
Thanks for review.

rimas-kudelis commented:
Wow, thanks for giving this a try! I'm pretty sure TTK has become even better now. :)
